core/vm/environment.go

```diff
@@ -17,39 +17,86 @@
 package vm
 
 import (
-	"errors"
 	"math/big"
 
 	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethereum/go-ethereum/core/state"
 )
 
 // Environment is is required by the virtual machine to get information from
-// it's own isolated environment. For an example see `core.VMEnv`
-type Environment interface {
-	State() *state.StateDB
+// it's own isolated environment.
 
+// Environment is an EVM requirement and helper which allows access to outside
+// information such like states.
+type Environment interface {
+	// The state database
+	Db() Database
+	// Creates a restorable snapshot
+	MakeSnapshot() Database
+	// Set database to previous snapshot
+	SetSnapshot(Database)
+	// Address of the original invoker (first occurance of the VM invoker)
 	Origin() common.Address
+	// The block number this VM is invoken on
 	BlockNumber() *big.Int
+	// The n'th hash ago from this block number
 	GetHash(n uint64) common.Hash
+	// The handler's address
 	Coinbase() common.Address
+	// The current time (block time)
 	Time() *big.Int
+	// Difficulty set on the current block
 	Difficulty() *big.Int
+	// The gas limit of the block
 	GasLimit() *big.Int
-	CanTransfer(from Account, balance *big.Int) bool
+	// Determines whether it's possible to transact
+	CanTransfer(from common.Address, balance *big.Int) bool
+	// Transfer from to to with amount set
```

Comment:
should be either "from from to to" or rephrased :) perhaps "transfer amount between from and to"